### PR TITLE
Offline grades

### PR DESCRIPTION
**Pull request by vargaat**
refs: MBL-16765
affects: Student
release note: none

test plan:
- Sync the grades tab for offline usage.
- Turn off internet on the device.
- Go to grades page.
- Page content should display the current grading period and grades should match the online version.
- Grading periods filter is disabled in offline mode.
- If grading period filter is modified in online mode then it will corrupt the offline data and it won't display the total grade, display assignments for the period selected in online mode instead of the last one while displaying the title of the last grading period.

## Checklist

- [x] Tested on phone
- [x] Tested on tablet

### CODE REVIEW

**Comment by inst-danger** [comment]
<details><summary>Student Build QR Code:</summary>

</details>

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>This pull request will not generate a release note.</td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>One or more files are below the minimum test coverage 50%</td>
    </tr>
  </tbody>
</table>


#### Affected Apps: Student

[MBL-16765](https://instructure.atlassian.net/browse/MBL-16765)

Coverage | New % | Master % | Delta
-------- | ----- | -------- | -----
**Canvas iOS** | 92.29% | 92.33% | -0.04%
Core/Core/Offline/ViewModel/OfflineModeViewModel.swift | 0% | 0% | 0%
Core/Core/Extensions/BoolExtensions.swift | 0% | 0% | 0%
Core/Core/Offline/View/NotSupportedInOfflineAlert.swift | 0% | -- | --
Core/Core/CourseSync/CourseSyncProgress/Model/CourseSyncProgressInteractor.swift | 0% | 0% | 0%
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 536dea893630ce0a6fc9fad9d86f848025ae2679
</p>

File: Core/Core/Grades/GradeListViewController.swift [hunk 1]
```diff
@@ -134,6 +142,17 @@ public class GradeListViewController: ScreenViewTrackableViewController, Colored
         navigationController?.navigationBar.useContextColor(color)
     }
 
+    public override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+
+        // Without this there was some weird empty space at the end of the tableview
+        // that went away after rotation or when we moved away from the screen and returned
+        if OfflineModeInteractorLive.shared.isOfflineModeEnabled() {
```
Comment (szabinst) [inline Core/Core/Grades/GradeListViewController.swift:150]: Could we inject the instance when we create the view controller?

File: Core/Core/Offline/View/UIButtonOfflineSupport.swift [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    // MARK: - Private Methods
+
+    private struct AssociatedObjectKeys {
+        static var OfflineStateObservation = "OfflineStateObservation"
+        static var TapGestureRecognizer = "TapGestureRecognizer"
+    }
+
+    private func makeUnavailable() {
```
Comment (szabinst) [inline Core/Core/Offline/View/UIButtonOfflineSupport.swift:47]: I'm not sure if this function name is expressive enough. I'm thinking something like `setDisabledState` or `setUnavailableState`.

File: Core/Core/Offline/View/UIButtonOfflineSupport.swift [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            return
+        }
+
+        let tapRecognizer = UITapGestureRecognizer(target: UIAlertController.self,
+                                                   action: #selector(UIAlertController.showItemNotAvailableInOfflineAlert))
+        addGestureRecognizer(tapRecognizer)
+        objc_setAssociatedObject(self, &AssociatedObjectKeys.TapGestureRecognizer, tapRecognizer, .OBJC_ASSOCIATION_RETAIN)
+    }
+
+    private func makeAvailable() {
```
Comment (szabinst) [inline Core/Core/Offline/View/UIButtonOfflineSupport.swift:63]: Same as above.

**Review by szabinst — DISMISSED** [review]
QA + 1

File: Core/Core/Grades/GradeListViewController.swift (hunk 1 above)
Comment (vargaat) [reply Core/Core/Grades/GradeListViewController.swift]: Yes, I modified it.

File: Core/Core/Offline/View/UIButtonOfflineSupport.swift (hunk 2 above)
Comment (vargaat) [reply Core/Core/Offline/View/UIButtonOfflineSupport.swift:47]: Agree, methods starting with make used to create and return something.

**Review by szabinst — APPROVED** [review]

**Review by szabinst — APPROVED** [review]

**Review by idevainstr — APPROVED** [review]
QA +1

**Review by balintbartok — APPROVED** [review]

**Pull request merged**